File: packages/python/onnx9000-tflite-exporter/src/onnx9000/tflite_exporter/flatbuffer/reader.py
import struct
from typing import Optional


class FlatBufferReader:
    """Zero-dependency FlatBuffer Reader."""

    def __init__(self, data: bytes) -> None:
        """Initialize the reader."""
        self.bytes = data

    def get_root(self) -> int:
        """Get the root object offset."""
        return struct.unpack_from("<I", self.bytes, 0)[0]

    def check_magic_bytes(self, magic: str) -> bool:
        """Check if magic bytes are present."""
        if len(self.bytes) < 8:
            return False
        return self.bytes[4:8].decode("ascii") == magic

    def get_vtable(self, obj_offset: int) -> int:
        """Get the vtable offset."""
        vtable_offset = struct.unpack_from("<i", self.bytes, obj_offset)[0]
        return obj_offset - vtable_offset

    def get_field_offset(self, obj_offset: int, vtable_index: int) -> int:
        """Get the field offset relative to obj_offset."""
        vtable = self.get_vtable(obj_offset)
        vtable_size = struct.unpack_from("<H", self.bytes, vtable)[0]
        field_offset = (vtable_index + 2) * 2

        if field_offset >= vtable_size:
            return 0

        offset_in_object = struct.unpack_from("<H", self.bytes, vtable + field_offset)[0]
        if offset_in_object == 0:
            return 0

        return obj_offset + offset_in_object

    def get_int8(self, obj_offset: int, vtable_index: int, def_val: int = 0) -> int:
        """Get int8 field."""
        offset = self.get_field_offset(obj_offset, vtable_index)
        return struct.unpack_from("<b", self.bytes, offset)[0] if offset != 0 else def_val

    def get_int16(self, obj_offset: int, vtable_index: int, def_val: int = 0) -> int:
        """Get int16 field."""
        offset = self.get_field_offset(obj_offset, vtable_index)
        return struct.unpack_from("<h", self.bytes, offset)[0] if offset != 0 else def_val

    def get_int32(self, obj_offset: int, vtable_index: int, def_val: int = 0) -> int:
        """Get int32 field."""
        offset = self.get_field_offset(obj_offset, vtable_index)
        return struct.unpack_from("<i", self.bytes, offset)[0] if offset != 0 else def_val

    def get_float32(self, obj_offset: int, vtable_index: int, def_val: float = 0.0) -> float:
        """Get float32 field."""
        offset = self.get_field_offset(obj_offset, vtable_index)
        return struct.unpack_from("<f", self.bytes, offset)[0] if offset != 0 else def_val

    def get_string(self, obj_offset: int, vtable_index: int) -> Optional[str]:
        """Get string field."""
        offset = self.get_field_offset(obj_offset, vtable_index)
        if offset == 0:
            return None
        offset += struct.unpack_from("<I", self.bytes, offset)[0]
        length = struct.unpack_from("<I", self.bytes, offset)[0]
        return self.bytes[offset + 4 : offset + 4 + length].decode("utf-8")

    def get_vector_length(self, vector_offset: int) -> int:
        """Get vector length."""
        return struct.unpack_from("<I", self.bytes, vector_offset)[0]

    def get_indirect_offset(self, obj_offset: int, vtable_index: int) -> int:
        """Get indirect object offset."""
        offset = self.get_field_offset(obj_offset, vtable_index)
        if offset == 0:
            return 0
        return offset + struct.unpack_from("<I", self.bytes, offset)[0]

File: packages/js/tflite-exporter/src/flatbuffer/reader.ts
/**
 * Zero-dependency FlatBuffer Reader.
 */
export class FlatBufferReader {
  private view: DataView;
  private bytes: Uint8Array;

  constructor(data: Uint8Array) {
    this.bytes = data;
    this.view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  }

  public getRoot(): number {
    const offset = this.view.getUint32(0, true);
    return offset;
  }

  public getRootAsModel(): number {
    // skip 4 bytes offset, skip identifier if present (which it is for TFL3)
    // Actually the root offset is just at byte 0.
    return this.view.getUint32(0, true);
  }

  public checkMagicBytes(magic: string): boolean {
    if (this.bytes.length < 8) return false;
    let actual = '';
    for (let i = 4; i < 8; i++) {
      actual += String.fromCharCode(this.bytes[i]!);
    }
    return actual === magic;
  }

  public getVTable(objOffset: number): number {
    return objOffset - this.view.getInt32(objOffset, true);
  }

  public getFieldOffset(objOffset: number, vtableIndex: number): number {
    const vtable = this.getVTable(objOffset);
    const vtableSize = this.view.getUint16(vtable, true);
    const fieldOffset = (vtableIndex + 2) * 2;

    if (fieldOffset >= vtableSize) return 0;

    const offsetInObject = this.view.getUint16(vtable + fieldOffset, true);
    if (offsetInObject === 0) return 0;

    return objOffset + offsetInObject;
  }

  public getInt8(objOffset: number, vtableIndex: number, def: number = 0): number {
    const offset = this.getFieldOffset(objOffset, vtableIndex);
    return offset !== 0 ? this.view.getInt8(offset) : def;
  }

  public getInt16(objOffset: number, vtableIndex: number, def: number = 0): number {
    const offset = this.getFieldOffset(objOffset, vtableIndex);
    return offset !== 0 ? this.view.getInt16(offset, true) : def;
  }

  public getInt32(objOffset: number, vtableIndex: number, def: number = 0): number {
    const offset = this.getFieldOffset(objOffset, vtableIndex);
    return offset !== 0 ? this.view.getInt32(offset, true) : def;
  }

  public getFloat32(objOffset: number, vtableIndex: number, def: number = 0.0): number {
    const offset = this.getFieldOffset(objOffset, vtableIndex);
    return offset !== 0 ? this.view.getFloat32(offset, true) : def;
  }

  public getVectorLength(vectorOffset: number): number {
    return this.view.getUint32(vectorOffset, true);
  }

  public getVectorItemOffset(vectorOffset: number, index: number, elemSize: number = 4): number {
    // 4 bytes length, followed by data
    return vectorOffset + 4 + index * elemSize;
  }

  public getString(objOffset: number, vtableIndex: number): string | null {
    let offset = this.getFieldOffset(objOffset, vtableIndex);
    if (offset === 0) return null;

    offset += this.view.getUint32(offset, true); // dereference string offset
    const length = this.view.getUint32(offset, true);

    let str = '';
    for (let i = 0; i < length; i++) {
      str += String.fromCharCode(this.bytes[offset + 4 + i]!);
    }
    return str;
  }

  public getIndirectOffset(objOffset: number, vtableIndex: number): number {
    const offset = this.getFieldOffset(objOffset, vtableIndex);
    if (offset === 0) return 0;
    return offset + this.view.getUint32(offset, true);
  }
}
